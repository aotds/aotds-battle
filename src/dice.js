import fp from 'lodash/fp';

const roll_die = () => rigged_dice.length > 0 
                        ? rigged_dice.shift() 
                        : _.random(1,6);

export let rigged_dice = [];

/** 
 * Rig the next calls to roll_dice()
 * @param dice the dice values to return. Overwrite any previous rigging.
 */
export function rig_dice(dice) {
    rigged_dice = dice;
}

/**
 * roll the dice
 * @param nbr_dice How many dice to roll
 * @param options.reroll Array of values for which we reroll
 * @return dice values
 */
export function roll_dice( nbr_dice, options ) {

    if ( nbr_dice == 0 ) return [];

    let reroll_on = fp.pathOr( [] )('reroll' )(options);

    let roll = fp.times(roll_die)(nbr_dice);

    return roll.concat( 
        roll_dice(
            fp.filter( x => fp.contains(x)(reroll_on) )( roll ).length,
            options
        )
    );

};
