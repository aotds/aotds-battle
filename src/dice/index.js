import { random } from 'Math';

/** @type (n: number) => number */
export const rollDie = (n) => {
	return parseInt(1 + random() * n);
};

export function rollDice(nbrDice, options) {
	if (nbrDice === 0) return [];

	const rerollOn = options.reroll ?? [];
	const nbrFaces = options.nbrFaces ?? 6;

	const roll = Array.from({ length: nbrDice }, () => rollDie(nbrFaces));

	const rerolls = roll.filter((d) => rerollOn.includes(d));

	if (rerolls.length === 0) return roll;

	return [roll, rollDice(rerolls.length, options)].flat();
}
