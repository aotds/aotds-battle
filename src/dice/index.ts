import  random  from 'lodash/random.js';

const rollDie = (n: number) => random(1, n);

export function rollDice(
	nbr_dice: number,
	options: { reroll?: number[]; nbr_faces?: number; note?: unknown } = {},
): number[] {
	if (nbr_dice === 0) return [];

	const reroll_on = options.reroll ?? [];
	const nbr_faces = options.nbr_faces ?? 6;

	const roll: number[] = Array.from({ length: nbr_dice }, () =>
		rollDie(nbr_faces),
	);

	const rerolls = roll.filter((d) => (reroll_on as any).includes(d));

	if (rerolls.length === 0) return roll;

	return ([roll, rollDice(rerolls.length, options)] as any).flat();
}
