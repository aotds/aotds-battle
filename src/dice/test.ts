jest.mock('lodash');

import { rollDice } from '.';
import { random } from 'lodash';

test('dice', () => {
	[6, 5, 4].forEach((x) => ( random  as any).mockReturnValueOnce(x));

	expect(rollDice(2, { reroll: [6] })).toEqual([6, 5, 4]);

	[6, 5, 6, 3].forEach((x) => ( random as any ).mockReturnValueOnce(x));
	expect(rollDice(2, { reroll: [6] })).toEqual([6, 5, 6, 3]);
});

test('nbr_faces', () => {
	( random as any ).mockImplementation((a, b) => b);
	expect(rollDice(1, { nbr_faces: 99 })).toEqual([99]);
});
