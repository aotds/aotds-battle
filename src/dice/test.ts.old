import tap from 'tap';
import {rollDice} from '.';
import _ from 'lodash';

let rigged_dice: number[] = [];

_.random = () => (rigged_dice as any).shift();

rigged_dice = [6, 5, 4];

tap.match(rollDice(2, {reroll: [6]}), [
    6, 5, 4
]);

rigged_dice = [6, 5, 6, 3];

tap.match(rollDice(2, {reroll: [6]}), [
    6, 5, 6, 3
]);

tap.test( 'nbr_faces', async () => {
    const original = _.random;
    _.random = ((one,n) => n ) as any;

    tap.match( rollDice(1,{nbr_faces: 99}), [ 99 ], 'got 99' );

    _.random = original;
});
