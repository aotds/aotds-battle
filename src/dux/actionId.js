import { action, Updux } from 'updux';
import u from 'updeep';

export const dux = new Updux({
	initial: 1,
	selectors: {
		getNextActionId: (state) => state,
	},
	actions: {
		incActionId: action(
			'incActionId',
			() => null,
			u.updateIn('meta.noLog', true),
		),
		setActionId: action(
			'setActionId',
			(id) => id,
			u.updateIn('meta.noLog', true),
		),
	},
	mutations: {
		incActionId: () => (x) => x + 1,
		setActionId: (id) => () => id,
	},
});

export const addActionIdEffect = (api) => (next) => (action) => {
	if (action.meta?.noLog) return next(action);

	// we don't want an infinite loop, do we?
	if (action.type === 'incActionId') return next(action);

	const actionId = api.getState.getNextActionId();
	api.dispatch(dux.actions.incActionId());

	return next(u({ meta: { actionId } }, action));
};
//dux.addEffect( '*', addActionId );

export const middlewareWrapper = (middlewares) => [addActionId, ...middlewares];
