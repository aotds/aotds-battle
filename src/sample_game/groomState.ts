import u from 'updeep';

expect.addSnapshotSerializer( {
  serialize(val, config, indentation, depth, refs, printer) {
    return JSON.stringify(groomState(val),null,2);
  },

  test(val) {
    return val && val.hasOwnProperty('game');
  },
}
)

function groomState(state) {
    return u({
        game: {
            next_action_id: u.omitted,
        },
        log: u.map(
            u.omit('meta')
        )
    })(state);
}
