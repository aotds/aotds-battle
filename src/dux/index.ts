import Updux from 'updux';
import fp from 'lodash/fp';
import u from 'updeep';
import { action, empty } from 'ts-action';

import { metaTimestampDux } from './metaTimestamp';
import actionId, { actionIdEffect } from './game/actionId';
import subactions from './subactions';
import playPhases from './playPhases';

const dux = new Updux({
    coduxes: [metaTimestampDux, playPhases],
    subduxes: { 'game.next_action_id': actionId },
    effects: [['*', actionIdEffect(state => state?.game?.next_action_id)]],
});
export default dux;

