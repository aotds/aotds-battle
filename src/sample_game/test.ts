/* eslint-disable @typescript-eslint/no-var-requires */

import { range } from 'lodash';
import { applyMiddleware } from 'redux';
import { composeWithDevTools } from 'remote-redux-devtools';

import { dux } from '../dux';

expect.addSnapshotSerializer({
	test: () => true,
	print: (val) => JSON.stringify(val, null, 2),
});

const composeEnhancers = composeWithDevTools({
	realtime: true,
	port: 8000,
	maxAge: 3000,
	actionsBlacklist: ['incActionId'],
});

const battle = dux.createStore(undefined, (middleware) => {
	if (process.env.REDUX_DEBUG)
		return composeEnhancers(applyMiddleware(middleware));

	return applyMiddleware(middleware);
});

const playTurn = async (round) => {
	const turn = require(`./turn-${round}`);

	// rigged_dice = turn.dice ?? [];

	(turn.actions ?? []).forEach(battle.dispatch);

	const state = battle.getState();

	expect(state).toMatchSnapshot();

	await turn.tests(state);
};

test.each(range(2))('turn %#', playTurn);

/*
import './groomState';

import battle_dux from '../dux';

jest.mock('../dice');
import * as dice from '../dice';

let rigged_dice = [];
dice.rollDice.mockImplementation((...args: any[]) => {
    if (!rigged_dice.length) {
        throw new Error(`no dice left for ${JSON.stringify(args)}`);
    }

    const index = fp.findIndex((d: any) => {
        if (Array.isArray(d)) return true;
        return fp.matches(d.match, {
            dice: args[0],
            ...args[1],
        });
    })(rigged_dice);

    if (index === -1) {
        throw new Error(`no dice found for ${JSON.stringify(args)}`);
    }

    const [result] = rigged_dice.splice(index, 1);

    return Array.isArray(result) ? result : result.dice;
});


*/
