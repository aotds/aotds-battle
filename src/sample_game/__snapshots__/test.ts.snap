// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`turn 0 1`] = `
{
  "actionId": {
    "nextActionId": 1
  },
  "game": {
    "name": "",
    "turn": 0
  },
  "bogeys": {},
  "log": []
}
`;

exports[`turn 1 1`] = `
{
  "actionId": {
    "nextActionId": 12
  },
  "game": {
    "name": "gemini",
    "turn": 1,
    "players": [
      "yanick",
      "yenzie"
    ]
  },
  "bogeys": {
    "enkidu": {
      "id": "enkidu",
      "name": "enkidu",
      "drive": {
        "current": 6,
        "rating": 6,
        "damageLevel": 0
      },
      "navigation": {
        "thrustUsed": 3,
        "trajectory": [
          {
            "type": "POSITION",
            "coords": [
              0,
              0
            ]
          },
          {
            "type": "BANK",
            "delta": [
              1,
              0
            ],
            "coords": [
              1,
              0
            ]
          },
          {
            "type": "ROTATE",
            "delta": 0,
            "heading": 0
          },
          {
            "type": "ROTATE",
            "delta": 1,
            "heading": 1
          },
          {
            "type": "MOVE",
            "delta": [
              0.5,
              0.87
            ],
            "coords": [
              1.5,
              0.87
            ]
          }
        ],
        "maneuvers": {
          "thrust": [
            0,
            4
          ],
          "bank": [
            -3,
            3
          ],
          "turn": [
            -3,
            3
          ]
        },
        "heading": 1,
        "coords": [
          1.5,
          0.87
        ],
        "velocity": 1,
        "effectiveOrders": {
          "thrust": 1,
          "bank": 1,
          "turn": 1
        }
      },
      "weaponry": {
        "firecons": {
          "1": {
            "id": 1
          }
        },
        "shields": {
          "1": {
            "id": 1,
            "level": 1,
            "damaged": false
          },
          "2": {
            "id": 2,
            "level": 2,
            "damaged": false
          }
        },
        "weapons": {
          "1": {
            "weaponType": "beam",
            "weaponClass": 2,
            "arcs": [
              "F"
            ],
            "id": 1
          },
          "2": {
            "arcs": [
              "FS"
            ],
            "weaponType": "beam",
            "weaponClass": 1,
            "id": 2
          },
          "3": {
            "arcs": [
              "FP"
            ],
            "weaponType": "beam",
            "weaponClass": 1,
            "id": 3
          }
        }
      },
      "structure": {
        "hull": {
          "rating": 4,
          "current": 4
        },
        "armor": {
          "rating": 4,
          "current": 4
        }
      },
      "playerId": "yanick",
      "orders": {
        "navigation": {
          "thrust": 1,
          "turn": 1,
          "bank": 1
        }
      }
    },
    "siduri": {
      "id": "siduri",
      "name": "siduri",
      "drive": {
        "rating": 6,
        "current": 6
      },
      "navigation": {
        "thrustUsed": 1,
        "trajectory": [
          {
            "type": "POSITION",
            "coords": [
              10,
              10
            ]
          },
          {
            "type": "MOVE",
            "delta": [
              0,
              -1
            ],
            "coords": [
              10,
              9
            ]
          }
        ],
        "maneuvers": {
          "thrust": [
            0,
            6
          ],
          "bank": [
            -3,
            3
          ],
          "turn": [
            -3,
            3
          ]
        },
        "heading": 6,
        "coords": [
          10,
          9
        ],
        "velocity": 1,
        "effectiveOrders": {
          "thrust": 1,
          "bank": 0,
          "turn": 0
        }
      },
      "playerId": "yenzie",
      "weaponry": {
        "shields": {
          "1": {
            "id": 1,
            "level": 1,
            "damaged": false
          },
          "2": {
            "id": 2,
            "level": 2,
            "damaged": false
          }
        },
        "firecons": {},
        "weapons": {}
      },
      "structure": {
        "hull": {
          "rating": 4,
          "current": 4
        },
        "shields": [
          1,
          2
        ],
        "armor": {
          "rating": 4,
          "current": 4
        }
      },
      "orders": {
        "navigation": {
          "thrust": 1,
          "bank": 0,
          "turn": 0
        }
      }
    }
  },
  "log": [
    {
      "type": "initGame",
      "payload": {
        "game": {
          "name": "gemini",
          "players": [
            "yanick",
            "yenzie"
          ]
        },
        "bogeys": [
          {
            "name": "enkidu",
            "drive": 6,
            "navigation": {
              "coords": [
                0,
                0
              ]
            },
            "weaponry": {
              "firecons": 1,
              "shields": [
                1,
                2
              ],
              "weapons": [
                {
                  "weaponType": "beam",
                  "weaponClass": 2,
                  "arcs": [
                    "F"
                  ]
                },
                {
                  "arcs": [
                    "FS"
                  ],
                  "weaponType": "beam",
                  "weaponClass": 1
                },
                {
                  "arcs": [
                    "FP"
                  ],
                  "weaponType": "beam",
                  "weaponClass": 1
                }
              ]
            },
            "structure": {
              "hull": 4,
              "armor": 4
            },
            "playerId": "yanick"
          },
          {
            "name": "siduri",
            "drive": {
              "rating": 6,
              "current": 6
            },
            "navigation": {
              "coords": [
                10,
                10
              ],
              "heading": 6
            },
            "playerId": "yenzie",
            "weaponry": {
              "shields": [
                1,
                2
              ]
            },
            "structure": {
              "hull": 4,
              "shields": [
                1,
                2
              ],
              "armor": 4
            }
          }
        ]
      },
      "meta": {
        "actionId": 1
      }
    },
    {
      "type": "setOrders",
      "payload": {
        "bogeyId": "enkidu",
        "orders": {
          "navigation": {
            "thrust": 1,
            "turn": 1,
            "bank": 1
          }
        }
      },
      "meta": {
        "actionId": 2
      }
    },
    {
      "type": "setOrders",
      "payload": {
        "bogeyId": "siduri",
        "orders": {
          "navigation": {
            "thrust": 1,
            "bank": 0,
            "turn": 0
          }
        }
      },
      "meta": {
        "actionId": 3
      }
    },
    {
      "type": "playTurn",
      "meta": {
        "actionId": 4
      }
    },
    {
      "type": "movementPhase",
      "meta": {
        "parentActionId": 4,
        "actionId": 5
      }
    },
    {
      "type": "bogeyMovementResolution",
      "payload": {
        "bogeyId": "enkidu",
        "movement": {
          "thrustUsed": 3,
          "trajectory": [
            {
              "type": "POSITION",
              "coords": [
                0,
                0
              ]
            },
            {
              "type": "BANK",
              "delta": [
                1,
                0
              ],
              "coords": [
                1,
                0
              ]
            },
            {
              "type": "ROTATE",
              "delta": 0,
              "heading": 0
            },
            {
              "type": "ROTATE",
              "delta": 1,
              "heading": 1
            },
            {
              "type": "MOVE",
              "delta": [
                0.5,
                0.87
              ],
              "coords": [
                1.5,
                0.87
              ]
            }
          ],
          "maneuvers": {
            "thrust": [
              0,
              4
            ],
            "bank": [
              -3,
              3
            ],
            "turn": [
              -3,
              3
            ]
          },
          "heading": 1,
          "coords": [
            1.5,
            0.87
          ],
          "velocity": 1,
          "effectiveOrders": {
            "thrust": 1,
            "bank": 1,
            "turn": 1
          }
        }
      },
      "meta": {
        "parentActionId": 5,
        "actionId": 6
      }
    },
    {
      "type": "bogeyMovementResolution",
      "payload": {
        "bogeyId": "siduri",
        "movement": {
          "thrustUsed": 1,
          "trajectory": [
            {
              "type": "POSITION",
              "coords": [
                10,
                10
              ]
            },
            {
              "type": "MOVE",
              "delta": [
                0,
                -1
              ],
              "coords": [
                10,
                9
              ]
            }
          ],
          "maneuvers": {
            "thrust": [
              0,
              6
            ],
            "bank": [
              -3,
              3
            ],
            "turn": [
              -3,
              3
            ]
          },
          "heading": 6,
          "coords": [
            10,
            9
          ],
          "velocity": 1,
          "effectiveOrders": {
            "thrust": 1,
            "bank": 0,
            "turn": 0
          }
        }
      },
      "meta": {
        "parentActionId": 5,
        "actionId": 7
      }
    },
    {
      "type": "fireconOrdersPhase",
      "meta": {
        "parentActionId": 4,
        "actionId": 8
      }
    },
    {
      "type": "weaponOrdersPhase",
      "meta": {
        "parentActionId": 4,
        "actionId": 9
      }
    },
    {
      "type": "weaponFiringPhase",
      "meta": {
        "parentActionId": 4,
        "actionId": 10
      }
    },
    {
      "type": "clearOrders",
      "meta": {
        "parentActionId": 4,
        "actionId": 11
      }
    }
  ]
}
`;
